File: python/858.py

import sys

# Increase recursion depth for deep recursion if necessary, though 9 is small.
sys.setrecursionlimit(2000)

def solve():
    MOD = 10**9 + 7
    N = 800
    
    # For testing with G(5)
    # N = 5
    
    # Primes and classification
    primes = []
    is_prime = [True] * (N + 1)
    is_prime[0] = is_prime[1] = False
    for p in range(2, N + 1):
        if is_prime[p]:
            primes.append(p)
            for i in range(p * p, N + 1, p):
                is_prime[i] = False
                
    sqrt_N = int(N**0.5)
    small_primes = [p for p in primes if p <= sqrt_N]
    large_primes = [p for p in primes if p > sqrt_N]
    
    # Generate smooth numbers (only small prime factors)
    smooths = []
    def generate_smooth(idx, current):
        if idx == len(small_primes):
            if current <= N:
                smooths.append(current)
            return

        p = small_primes[idx]
        # Try all powers of p
        pe = 1
        while current * pe <= N:
            generate_smooth(idx + 1, current * pe)
            pe *= p
    
    generate_smooth(0, 1)
    smooths.sort()
    num_smooths = len(smooths)
    
    # Map from value to index in smooths
    # Actually we only need to look up counts.
    # We need to find index i such that smooths[i] <= K < smooths[i+1]
    # bisect_right
    from bisect import bisect_right
    
    # Constants W_p for all primes
    # W_p = p^k_max where p^k_max <= N < p^(k_max+1)
    W_vals = {}
    for p in primes:
        pe = 1
        while pe * p <= N:
            pe *= p
        W_vals[p] = pe
        
    # Precompute constant C = product of W_p
    C = 1
    for p in primes:
        C = (C * W_vals[p]) % MOD
        
    # Modular Inverse
    def modinv(a):
        return pow(a, MOD - 2, MOD)
        
    # Weight function w(p, k)
    # if k=0: 1
    # if k>0: -phi(p^k) / W_p
    def get_weight(p, k):
        if k == 0:
            return 1
        # phi(p^k) = p^k - p^(k-1)
        phi = (pow(p, k, MOD) - pow(p, k-1, MOD)) % MOD
        num = (-phi) % MOD
        den = W_vals[p]
        return (num * modinv(den)) % MOD

    # Groups for large primes
    # Map K -> List of large primes P such that floor(N/P) = K
    # Only need K where list is non-empty.
    # K = N // P.
    large_groups = {}
    for P in large_primes:
        K = N // P
        if K not in large_groups:
            large_groups[K] = []
        large_groups[K].append(P)
        
    # Precompute polynomials for each group
    # Poly_K[cnt] = Product_{P in group} (2^cnt + w(P, 1))
    # w(P, 1) since P is large, max exponent is 1.
    # w(P, 1) = -(P-1)/P (mod MOD) since W_P = P for P > sqrt(N) (mostly)
    # Wait, if P > sqrt(N), then P^2 > N, so W_P = P. Correct.
    
    # We need table Lookup_K[cnt] for cnt in 0..idx_K
    # idx_K is count of smooths <= K.
    
    group_lookups = [] # List of (idx_mask, lookup_table)
    
    for K, P_list in large_groups.items():
        limit_idx = bisect_right(smooths, K)
        # Lookup table size limit_idx + 1
        table = [1] * (limit_idx + 1)
        
        # Compute product
        # For each P, compute term = (2^cnt + w(P))
        # w(P) = -(P-1)/P
        # But we can sum log-probs? No, sum over field.
        
        # Calculate w_P list
        ws = []
        for P in P_list:
            w = get_weight(P, 1)
            ws.append(w)
            
        # Populate table
        # For each possible cnt
        # This might be slow if group is large and table is large?
        # Group size can be large (e.g. K=1, many P).
        # But table is small (limit_idx for K=1 is 1, smooths <= 1 is {1}).
        # For K=27, limit_idx is larger.
        
        for cnt in range(limit_idx + 1):
            val = 1
            pow2 = pow(2, cnt, MOD)
            for w in ws:
                term = (pow2 + w) % MOD
                val = (val * term) % MOD
            table[cnt] = val
            
        # Determine mask for this K
        # We need mask to extract bits 0..limit_idx-1
        # mask = (1 << limit_idx) - 1
        # Actually we'll store limit_idx
        group_lookups.append({
            'limit': limit_idx,
            'table': table
        })
        
    # Also need the term for the smooth numbers themselves (K=N)
    # This corresponds to the factor 2^M(cS) * W(cS)
    # M(cS) is bit_count of full mask.
    # We handle W(cS) in the recursion.
    # The 2^M(cS) is simply pow(2, bit_count, MOD).
    # We can precompute powers of 2.
    pow2_all = [pow(2, i, MOD) for i in range(num_smooths + 1)]

    # Precompute bitmasks for small primes
    # masks[p][k] = integer bitmask
    masks = {}
    
    for p in small_primes:
        masks[p] = {}
        # Possible exponents: 0 to max_e
        max_e = 0
        pe = 1
        while pe <= N:
            pe *= p
            max_e += 1
        max_e -= 1 # Correct
        
        # k=0
        # fits if v_p(y) < infinity (always)
        masks[p][0] = (1 << num_smooths) - 1
        
        for k in range(1, max_e + 1):
            m = 0
            for i, y in enumerate(smooths):
                # Check v_p(y)
                v = 0
                temp = y
                while temp > 0 and temp % p == 0:
                    v += 1
                    temp //= p
                
                if v < k:
                    m |= (1 << i)
            masks[p][k] = m

    # Recursion
    total_sum = 0
    
    # Prepare list of small primes and their options
    small_options = []
    for p in small_primes:
        opts = []
        # k=0
        opts.append((0, get_weight(p, 0)))
        # k=1..max_e
        pe = p
        k = 1
        while pe <= N:
            opts.append((k, get_weight(p, k)))
            pe *= p
            k += 1
        small_options.append((p, opts))
        
    def recurse(idx, current_mask, current_weight):
        nonlocal total_sum
        
        if idx == len(small_options):
            # Leaf
            term = current_weight
            
            # Factor from smooth numbers (2^M)
            M = current_mask.bit_count()
            term = (term * pow2_all[M]) % MOD
            
            # Factors from large groups
            for grp in group_lookups:
                limit = grp['limit']
                # extract bits relevant to this group
                # The first 'limit' bits of current_mask
                # mask is (1 << limit) - 1
                sub_mask = current_mask & ((1 << limit) - 1)
                cnt = sub_mask.bit_count()
                term = (term * grp['table'][cnt]) % MOD
                
            total_sum = (total_sum + term) % MOD
            return

        p, opts = small_options[idx]
        p_masks = masks[p]
        
        for k, w in opts:
            # Update mask
            new_mask = current_mask & p_masks[k]
            # Update weight
            new_weight = (current_weight * w) % MOD
            
            recurse(idx + 1, new_mask, new_weight)
            
    # Start recursion
    initial_mask = (1 << num_smooths) - 1
    recurse(0, initial_mask, 1)
    
    final_ans = (total_sum * C) % MOD
    return final_ans

if __name__ == "__main__":
    print(solve())

File: c/858.c
/*
 * Project Euler 858 - G(N) computation
 *
 * Computes G(800) using smooth numbers, prime weight recursion, and bitmask DP.
 * Smooth numbers (only small prime factors) with up to 322 entries require
 * multi-word bitmask handling.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MOD 1000000007LL
#define N 800

/* Bitmask with 6 * 64 = 384 bits, enough for 322 */
#define BM_WORDS 6
typedef struct { unsigned long long w[BM_WORDS]; } bitmask;

static bitmask bm_zero(void) {
    bitmask b; memset(&b, 0, sizeof(b)); return b;
}

static bitmask bm_full(int nbits) {
    bitmask b = bm_zero();
    for (int i = 0; i < nbits; i++)
        b.w[i / 64] |= 1ULL << (i % 64);
    return b;
}

static void bm_set(bitmask *b, int i) { b->w[i/64] |= 1ULL << (i%64); }
static int bm_test(const bitmask *b, int i) { return (b->w[i/64] >> (i%64)) & 1; }

static bitmask bm_and(bitmask a, bitmask b) {
    bitmask r;
    for (int i = 0; i < BM_WORDS; i++) r.w[i] = a.w[i] & b.w[i];
    return r;
}

static int bm_popcount(bitmask b) {
    int c = 0;
    for (int i = 0; i < BM_WORDS; i++)
        c += __builtin_popcountll(b.w[i]);
    return c;
}

/* Extract popcount of first 'lim' bits */
static int bm_popcount_first(bitmask b, int lim) {
    int c = 0;
    for (int i = 0; i < lim; i++)
        if (bm_test(&b, i)) c++;
    return c;
}

static long long mod_pow(long long base, long long exp, long long mod) {
    long long r = 1;
    base %= mod; if (base < 0) base += mod;
    while (exp > 0) {
        if (exp & 1) r = r * base % mod;
        base = base * base % mod;
        exp >>= 1;
    }
    return r;
}

static long long modinv(long long a) { return mod_pow(a, MOD - 2, MOD); }

/* Primes */
static int primes[200], nprimes;
static int small_primes[20], nsmall;
static int large_primes[200], nlarge;
static int is_prime_arr[N + 1];

/* Smooth numbers */
static int smooths[400];
static int num_smooths;

/* W_p values */
static long long W_vals[N + 1]; /* W_vals[p] for prime p */
static long long C_val;

/* Masks for small primes */
static bitmask sp_masks[20][20]; /* sp_masks[sp_idx][k] */
static int sp_max_e[20];

/* Options for small primes */
typedef struct { int k; long long weight; } option;
static option sp_options[20][20]; /* sp_options[sp_idx][opt_idx] */
static int sp_nopts[20];

/* Group lookups for large primes */
typedef struct {
    int limit; /* bisect_right(smooths, K) */
    long long table[400]; /* table[cnt] for cnt = 0..limit */
} group_lookup;
static group_lookup grp_lookups[200];
static int ngrp_lookups;

/* Powers of 2 */
static long long pow2_all[400];

/* Recursion */
static long long total_sum;

static void recurse(int idx, bitmask current_mask, long long current_weight) {
    if (idx == nsmall) {
        long long term = current_weight;

        /* Factor from smooth numbers (2^M) */
        int M = bm_popcount(current_mask);
        term = term * pow2_all[M] % MOD;

        /* Factors from large groups */
        for (int g = 0; g < ngrp_lookups; g++) {
            int lim = grp_lookups[g].limit;
            int cnt = bm_popcount_first(current_mask, lim);
            term = term * grp_lookups[g].table[cnt] % MOD;
        }

        total_sum = (total_sum + term) % MOD;
        return;
    }

    for (int oi = 0; oi < sp_nopts[idx]; oi++) {
        int k = sp_options[idx][oi].k;
        long long w = sp_options[idx][oi].weight;
        bitmask new_mask = bm_and(current_mask, sp_masks[idx][k]);
        long long new_weight = current_weight * w % MOD;
        recurse(idx + 1, new_mask, new_weight);
    }
}

static void generate_smooth(int idx, long long current) {
    if (idx == nsmall) {
        if (current <= N)
            smooths[num_smooths++] = (int)current;
        return;
    }
    int p = small_primes[idx];
    long long pe = 1;
    while (current * pe <= N) {
        generate_smooth(idx + 1, current * pe);
        pe *= p;
    }
}

static int bisect_right_smooths(int val) {
    int lo = 0, hi = num_smooths;
    while (lo < hi) {
        int mid = (lo + hi) / 2;
        if (smooths[mid] <= val) lo = mid + 1; else hi = mid;
    }
    return lo;
}

static int cmp_int(const void *a, const void *b) {
    return *(const int *)a - *(const int *)b;
}

int main(void) {
    /* Sieve */
    memset(is_prime_arr, 1, sizeof(is_prime_arr));
    is_prime_arr[0] = is_prime_arr[1] = 0;
    for (int p = 2; p * p <= N; p++)
        if (is_prime_arr[p])
            for (int i = p * p; i <= N; i += p)
                is_prime_arr[i] = 0;

    nprimes = nsmall = nlarge = 0;
    int sqrt_N = 28; /* floor(sqrt(800)) = 28 */
    for (int p = 2; p <= N; p++) {
        if (!is_prime_arr[p]) continue;
        primes[nprimes++] = p;
        if (p <= sqrt_N) small_primes[nsmall++] = p;
        else large_primes[nlarge++] = p;
    }

    /* Generate smooth numbers */
    num_smooths = 0;
    generate_smooth(0, 1);
    qsort(smooths, num_smooths, sizeof(int), cmp_int);

    /* W_vals */
    for (int i = 0; i < nprimes; i++) {
        int p = primes[i];
        long long pe = 1;
        while (pe * p <= N) pe *= p;
        W_vals[p] = pe;
    }

    /* C = product W_p mod MOD */
    C_val = 1;
    for (int i = 0; i < nprimes; i++)
        C_val = C_val * W_vals[primes[i]] % MOD;

    /* Weight function */
    /* get_weight(p, k): k==0 -> 1; else -(p^k - p^(k-1)) / W_p mod MOD */

    /* Prepare small prime options and masks */
    for (int si = 0; si < nsmall; si++) {
        int p = small_primes[si];

        /* max exponent */
        int max_e = 0;
        long long pe = 1;
        while (pe <= N) { pe *= p; max_e++; }
        max_e--;
        sp_max_e[si] = max_e;

        /* Masks */
        /* k=0: all bits set */
        sp_masks[si][0] = bm_full(num_smooths);

        for (int k = 1; k <= max_e; k++) {
            bitmask m = bm_zero();
            for (int i = 0; i < num_smooths; i++) {
                int y = smooths[i];
                int v = 0;
                int temp = y;
                while (temp > 0 && temp % p == 0) { v++; temp /= p; }
                if (v < k) bm_set(&m, i);
            }
            sp_masks[si][k] = m;
        }

        /* Options */
        sp_nopts[si] = 0;
        /* k=0 */
        sp_options[si][sp_nopts[si]].k = 0;
        sp_options[si][sp_nopts[si]].weight = 1;
        sp_nopts[si]++;

        pe = p;
        for (int k = 1; k <= max_e; k++) {
            long long phi = (mod_pow(p, k, MOD) - mod_pow(p, k - 1, MOD) + MOD) % MOD;
            long long num = (MOD - phi) % MOD;
            long long w = num * modinv(W_vals[p]) % MOD;
            sp_options[si][sp_nopts[si]].k = k;
            sp_options[si][sp_nopts[si]].weight = w;
            sp_nopts[si]++;
            pe *= p;
        }
    }

    /* Group lookups for large primes */
    /* Group by K = N // P */
    typedef struct { int K; int P_list[200]; int nP; } group_t;
    group_t groups[200];
    int ngroups = 0;

    /* Collect groups */
    int group_map[N + 1];
    memset(group_map, -1, sizeof(group_map));

    for (int i = 0; i < nlarge; i++) {
        int P = large_primes[i];
        int K = N / P;
        if (group_map[K] == -1) {
            group_map[K] = ngroups;
            groups[ngroups].K = K;
            groups[ngroups].nP = 0;
            ngroups++;
        }
        int gi = group_map[K];
        groups[gi].P_list[groups[gi].nP++] = P;
    }

    ngrp_lookups = ngroups;
    for (int gi = 0; gi < ngroups; gi++) {
        int K = groups[gi].K;
        int limit_idx = bisect_right_smooths(K);
        grp_lookups[gi].limit = limit_idx;

        /* Compute weights for each P in group */
        for (int cnt = 0; cnt <= limit_idx; cnt++) {
            long long val = 1;
            long long p2 = mod_pow(2, cnt, MOD);
            for (int pi = 0; pi < groups[gi].nP; pi++) {
                int P = groups[gi].P_list[pi];
                /* w(P, 1) = -(P-1) / W_P mod MOD */
                /* W_P = P for large primes */
                long long phi = (P - 1) % MOD;
                long long w = (MOD - phi) % MOD * modinv(W_vals[P]) % MOD;
                long long term = (p2 + w) % MOD;
                val = val * term % MOD;
            }
            grp_lookups[gi].table[cnt] = val;
        }
    }

    /* Powers of 2 */
    for (int i = 0; i <= num_smooths; i++)
        pow2_all[i] = mod_pow(2, i, MOD);

    /* Recurse */
    total_sum = 0;
    bitmask initial_mask = bm_full(num_smooths);
    recurse(0, initial_mask, 1);

    long long ans = total_sum * C_val % MOD;
    printf("%lld\n", ans);
    return 0;
}
